### ISSUE TITLE
Ticket Lock and Timezone problems for Australia/Brisbane after daylight saving change in AEST

### ISSUE THREAD

**Issue by perryk** [osTicket/osTicket#3991]
### Prerequisites

* [ Y] Can you reproduce the problem in a fresh installation of the "develop" branch?
* [ N] Do you have any errors in the PHP error log, or javascript console?
* [ Y] Did you check the [osTicket forums](http://osticket.com/forum/categories)?
* [ Y] Did you [perform a cursory search](https://github.com/osTicket/osTicket/issues) to see if your bug or enhancement is already reported?

### Versions

![screenshot from 2017-10-08 23-59-09](https://user-images.githubusercontent.com/806269/31323141-98d7a5fe-ace6-11e7-8aa3-ad27f910b775.png)

### Description

After daylight savings started in the AEST timezone, osTicket installations set to Brisbane time are unable to reply or edit tickets as the ticket lock renew drop down windows immediately appears and if renew is selected it just waits forever. If the tick lock is returned off, the system still displays all time-dates as being 1 hour behind the actual time in Brisbane. The system report incorrect reports the database timezone as being interpreted as Australia/Melbourne.

I've been around and around a few times on this, initially, I thought it started before the change to daylight saving for all the other state capitals in the AEST timezone. Brisbane is different in that it doesn't observe daylight savings whilst Melbourne and Sydney do.

I found a couple of PR's which both looked to fix the lock issue, however they didn't fix the wrong time issue. In particular [this](https://github.com/osTicket/osTicket/pull/3734) PR is interesting but it only fixes cases in which there are multiple timezones which might be returned, not the case I have here, although it looks similar.

In the end, I started working with a stock standard v1.10.1 branch of the code and found there were no ticket lock problems if the system date was 30 Sept, and there _were_ problems if the system date is 1 Oct or later. This matches with the start of daylight saving AEST and confirms I don't need the other PRs. This also confirmed that the system report was always reporting the timezone being interpreted as Australia/Melbourne.

Further complicating this I worked out was that I was using a database backup from Sept for my testing and there was session data in the ost_session database which included a cached variable for the timezone.

From what I gather there is initially a call to "SELECT @@system_time_zone.global;" to obtain the database timezone before the result is saved to the session data.

The select statement above returns "AEST" for me, however, this is then passed through the format class where the PHP function "timezone_name_from_abbr" is incorrectly returning "Australia/Melbourne" instead of "Australia/Brisbane" and saving this to the session data.

The timezone class has additional code for some interesting looking functions which are trying to work out ambiguities and has an array of years with daylight saving time etc, however, it looks like this code is only ever used if the database fails to report any type of timezone. The comment on the array is "Rules compiled from jstz rules.js". So this doesn't look to be used here.

I've just worked around by always returning "Australia/Brisbane" always which isn't ideal but works well enough for me. 

Thanks for any suggestions you may have on this.

Cheers.

### Steps to Reproduce

1. Change server and php timezone Australia/Brisbane. Database timezone can remain as "SYSTEM", changing it directly does not change the behaviour. The date should be after Oct and before Apr.
2. Attempt to post a reply or internal note.

**Expected behavior:** : To be able to type into the message box.

**Actual behavior:** : A ticket lock is expiring soon message appears which cannot be renewed.

**Comment by JediKev** [comment]
Yay the proper issue report! Lol

What do you get when you run the following in MySQL:
```
mysql> SELECT @@global.time_zone, @@session.time_zone;
```

**Comment by JediKev** [comment]
Also, have you gone through and [populated MySQL's time zone tables](https://dev.mysql.com/doc/refman/5.5/en/time-zone-support.html) and set an explicit time zone? (Ex. `Australia/Brisbane` instead of `AEST`)

**Comment by perryk** [comment]
Thanks JediKev.

MariaDB remains with the default of "SYSTEM".
![screenshot from 2017-10-09 13-09-37](https://user-images.githubusercontent.com/806269/31324154-20bb2fa2-acf3-11e7-9f19-5bf7710010d3.png)

Although I have tried with your good suggestion also. This is after populating the timezone tables and adding this line to /etc/my.cnf

[mysqld]
...
default-time-zone='Australia/Brisbane'

After setting this the 2 relevant mysql queries are:
![screenshot from 2017-10-09 13-15-47](https://user-images.githubusercontent.com/806269/31324228-fc307e16-acf3-11e7-9f19-6d544ec61d6d.png)

![screenshot from 2017-10-09 13-17-05](https://user-images.githubusercontent.com/806269/31324245-32c11224-acf4-11e7-9e1f-ff568ec7ca9e.png)
^ n.b the actual command I think include/mysqli.php is running is SELECT @@system_time_zone;. I have this incorrect in my issue above.

After rebooting and clearing out anything in ost_session etc and trying again I've still the same problem.

![screenshot from 2017-10-09 13-19-47](https://user-images.githubusercontent.com/806269/31324289-9eb7b0d2-acf4-11e7-854f-5bedad2f0d6e.png)

So AEST is returned even if I have MariaDB set to "SYSTEM" or "Australia/Brisbane" and the returned "AEST" is technically correct. It is just after that the "timezone_name_from_abbr" php function is returning Melbourne instead of Brisbane.

**Comment by clonemeagain** [comment]
I noticed that all my Agents had their profile's timezones set to Australia/Brisbane, but we are in Melbourne.. so, I just reset them all and it was back in action: 

```sql
UPDATE `ost_staff` set timezone = 'Australia/Sydney' WHERE
staff_id != 54 AND timezone != 'Australia/Sydney'
```

Obviously our 54'th Agent is actually in Brisbane, so that was fine.

Change to suit your timezone.. obviously only fixes an issue where the Agent's timezone is incorrect, you can't actually set it via the web interface as admin, each Agent has to configure it himself.

**Comment by perryk** [comment]
That is interesting, I'm really not sure what "timezone_name_from_abbr" is using to base it's decision on.

**Comment by JediKev** [comment]
`timezone_name_from_abbr` is using MySQL's `@@global.system_time_zone` variable to base it's decision on. It's crazy because I tried throwing `AEST` in that function with the `10` offset (for Brisbane) and it came out as `Australia/Melbourne`. So I tried every combination of things and nothing would give me `Australia/Brisbane`, only `Australia/Melbourne`. So I'm looking into what you have to input to get `Australia/Brisbane`. 👍

**Comment by perryk** [comment]
Thanks JediKev, I don't quite recall but I think it was returning Australia/Melbourne for me before the start of Oct. Prior to Oct, the time is the same in Brisbane and Melbourne so there was no ticket lock problem. Obviously, it was returning Australia/Brisbane at some point for Clonemeagain, however, perhaps he was running a different version of PHP or has changed PHP versions along the way.

**Comment by clonemeagain** [comment]
@perryk Likely mate, PHP 5.4 on that box.

**Comment by JediKev** [comment]
@perryk Could you try something for me? Try changing line `118` in `include/mysqli.php` from:
```
return db_get_variable('system_time_zone', 'global');
```
to:
```
return db_get_variable('time_zone', 'global');
```

Test with that and let me know if it returns your correct timezone and if your times are correct again.

**Comment by perryk** [comment]
First off, I'm sorry for the slow response, it has been busy here then when I sat down to look at this yesterday I found everything was working even without my workaround as my Vagrant setup with a timezone plugin was forcing a timezone to "Etc/GMT+10" which actually works also but is a little odd.

Today I started again without that in the Vagrant setup (and after confirming I still had the issue) I found changing that line 118 in mysqli.php as suggested fixes the problem nicely!! 

The output from system information is: 

![screenshot from 2017-10-24 21-51-32](https://user-images.githubusercontent.com/806269/31941546-90ad9a66-b905-11e7-961d-9536d1e42cc6.png)

Very interesting as I didn't think it would know what to do with the return value of "SYSTEM". 

I'll move to using your suggestion as the workaround, thanks heaps for looking into this.

**Comment by JediKev** [comment]
@perryk no need to explain yourself! I understand how crazy it can get (main reason why I've been slow in replying too). 😁 

That's awesome! I found out that PHP's `timezone_name_from_abbr` function apparently has an issue with AEST as a timezone (after trying everything I could to get Australia/Brisbane with no success). AEST is supposed to be Brisbane and AEDT is supposed to be Melbourne, right? Well to PHP's `timezone_name_from_abbr` AEST is still considered Melbourne. Why? Idk. Another crazy thing is, there is no way to change the `@@global.system_time_zone variable` variable. So if you live in Brisbane and your server timezone is AEST you're screwed. So, I'm going to talk about this with Peter and see if we should change the variable from `@@global.system_time_zone` to `@@global.time_zone`. This way if anyone is having timezone issues they can update their MySQL's `@@global.time_zone` variable to their correct timezone and PHP should interpret that correctly.

**Comment by perryk** [comment]
Correct, currently AEST is for Brisbane, and AEDT is for Melbourne. After daylight savings has finished AEST will cover Melbourne again though.

Yes, if something is amiss such as this or the incorrect timezone returned as per #3734 , then being able to manually specify it in the database to specify this sounds good.

I appreciate it might be something others may want to test out or take time to consider etc. The whole figuring out a timezone thing probably just isn't as simple as it might sound. I think I ended up down a few rabbit holes with this, but that was compounded by other issues and not really knowing all that much. I learnt a bit at least :)

For me, I'm happy with the workaround in place.

**Comment by QANTau** [comment]
I'm having similar time zone issues (located in Brisbane).

The server is Ubuntu 16.04 LTS, patches are up-to-date.  Server timezone is 'Australia\Brisbane' and everything seems OK:
![svr-tz](https://user-images.githubusercontent.com/1386766/37236976-4db748f0-2459-11e8-9c65-9ed9053dfa78.PNG)

php.ini contains:
`date.timezone = Australia/Brisbane`

MySQL timezone settings:
![mysql-tz](https://user-images.githubusercontent.com/1386766/37236930-c6206e4e-2458-11e8-9a87-c84be7a3cc56.PNG)

includes/mysqli.php has been modified as suggested:
![mysqli](https://user-images.githubusercontent.com/1386766/37237016-979519e8-2459-11e8-961f-af9eae7a7a7d.PNG)

My osTicket installation interprets the time zone as 'Australia/Melbourne':
![osticket](https://user-images.githubusercontent.com/1386766/37237019-a52de846-2459-11e8-825d-8951bbb405ee.PNG)

I've also set timezone 'Australia/Brisbane' manually in the system settings and on the user accounts, with no change to behaviour. 

All times are currently 1 hour out in the system, although this isn't causing any major issues for this particular (very small) installation, this is attached in case it can help with sorting this problem out.

> ... then being able to manually specify it in the database to specify this sounds good.

Agreed, just being able to override the default timezone handling would provide an effective solution for these corner-case timezone issues.

**Comment by JediKev** [comment]
@QANTau

Stupid question: You’ve restarted MySQL and Apache just to make sure?

**Comment by QANTau** [comment]
Yes, MySQL and Apache have been restarted.

Sorry ... should have already included that info :)

**Comment by perryk** [comment]
Hey @QANTau , it is nice to see someone else from Brissy.

I've taken a shot tonight at replicating your setup. Well, everything except I stuck with MariaDB instead of MySQL. I can change that pretty easy however as I have this all under a Vagrant/Ansible setup.

I'm not able to replicate your problem yet, however a couple of things to note from working on this before are:

Logout, then delete all (or just yours if others are using the instance) the entries from the sessions table between code changes or changing anything to do with timezones. The timezone calculated is saved in your session details and never re-calculated again for that session. 

Don't actually modify the database timezone, leave it as "SYSTEM", this seems to be the trick to get it to use the magic code in DbTimezone::determine - which is called from class.timezone.php.

Finally, I'll add one more stupid question just to support JediKev, You're ran `./manage deploy <webroot>` if your git repo is not in your webroot ?

**Comment by QANTau** [comment]
Apologies for the delay in replying, was travelling last week.

> Logout, then delete all (or just yours if others are using the instance) the entries from the sessions table between code changes or changing anything to do with timezones. The timezone calculated is saved in your session details and never re-calculated again for that session.

Although I didn't follow your instructions, no changes were made to the environment since I took the earlier screenshots but now the problem has resolved itself (after logging in again after a few days).  I'm guessing that I had the settings correct (at least as far as I can see) but my user session needed to be reset (which in my case was timed out).

![image](https://user-images.githubusercontent.com/1386766/37560239-60ba1f40-2a80-11e8-87d5-dc10619c5417.png)

> Finally, I'll add one more stupid question just to support JediKev, You're ran ./manage deploy <webroot> if your git repo is not in your webroot ?

No, my install is in webroot.

To summarise for anyone else struggling with Australia/Brisbane (once your server & application settings are OK) would need to clear their sessions as recommended by @perryk.

Thanks to everyone involved for the assist.

**Comment by perryk** [comment]
No troubles, good to hear all is working ok now.

**Comment by mjunaidahmed** [comment]
> @perryk Could you try something for me? Try changing line `118` in `include/mysqli.php` from:
> 
> ```
> return db_get_variable('system_time_zone', 'global');
> ```
> 
> to:
> 
> ```
> return db_get_variable('time_zone', 'global');
> ```
> 
> Test with that and let me know if it returns your correct timezone and if your times are correct again.

I had the exact same issue with V 1.17.4 on mySQL. This suggestion fixed the issue.

**Comment by perryk** [comment]
Yes, with that working for you I am sure it would work for me also.

There was a PR floating around for a while which I think did something similar. Likely PR #4186.

In more recent times we tend to work around the problem by running the following on the mysql server:

```echo 'default-time-zone = "+10:00"' >> /etc/my.cnf.d/mysql-server.cnf```
